File: sdk/go/ominiconnect/llm.go
package ominiconnect

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
)

// ─── LlmManager ───────────────────────────────────────────────────────────────

// LlmManager handles LLM-driven tool discovery and execution.
type LlmManager struct {
	client *Client
}

// Execute sends a natural language query and returns tool candidates or execution plan.
func (m *LlmManager) Execute(ctx context.Context, query string, platform *string) (*LlmExecuteResponse, error) {
	payload := map[string]any{"query": query}
	if platform != nil {
		payload["platform"] = *platform
	}
	resp, err := m.client.do(ctx, http.MethodPost, "/api/llm", nil, payload)
	if err != nil {
		return nil, err
	}
	if resp == nil {
		return nil, nil
	}
	var out LlmExecuteResponse
	if err := json.Unmarshal(*resp, &out); err != nil {
		return nil, fmt.Errorf("decode llm response: %w", err)
	}
	return &out, nil
}

// ListAvailableTools returns available tools per platform, filtered by platform if specified.
func (m *LlmManager) ListAvailableTools(ctx context.Context, platform *string) (*LlmToolsResponse, error) {
	query := url.Values{}
	if platform != nil {
		query.Set("platform", *platform)
	}
	resp, err := m.client.do(ctx, http.MethodGet, "/api/llm/tools", query, nil)
	if err != nil {
		return nil, err
	}
	if resp == nil {
		return nil, nil
	}
	var out LlmToolsResponse
	if err := json.Unmarshal(*resp, &out); err != nil {
		return nil, fmt.Errorf("decode llm tools: %w", err)
	}
	return &out, nil
}

// ─── Response types ──────────────────────────────────────────────────────────

type LlmExecuteResponse struct {
	OK                 bool             `json:"ok"`
	Tool               string           `json:"tool,omitempty"`
	ToolName           string           `json:"tool_name,omitempty"`
	Arguments          any              `json:"arguments,omitempty"`
	Explanation        string           `json:"explanation,omitempty"`
	Result             any              `json:"result,omitempty"`
	Error              string           `json:"error,omitempty"`
	Message            string           `json:"message,omitempty"`
	Candidates         []CandidateTool  `json:"candidates,omitempty"`
	AvailableToolsHint string           `json:"available_tools_hint,omitempty"`
}

type CandidateTool struct {
	Tool        string `json:"tool"`
	Name        string `json:"name"`
	MatchReason string `json:"match_reason"`
}

type LlmToolsResponse struct {
	Platforms map[string]PlatformTools `json:"platforms"`
}

type PlatformTools struct {
	Connected bool           `json:"connected"`
	Tools     []AvailableTool `json:"tools,omitempty"`
}

type AvailableTool struct {
	Slug           string   `json:"slug"`
	Name           string   `json:"name"`
	Description    string   `json:"description"`
	ExampleQueries []string `json:"example_queries"`
	Scopes         []string `json:"scopes"`
	ScopeSatisfied string   `json:"scope_satisfied"`
}

File: python/python/ominiconnect/llm.py
"""
LLM-powered tool execution API.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from .models import AvailableTool, LlmExecuteResponse, LlmToolsResponse


class LlmManager:
    """LLM-powered tool execution via OminiConnect.

    The LLM endpoint parses natural-language queries and routes them to
    appropriate tools, returning structured execution results.
    """

    def __init__(self, client: "OminiConnect"):
        self._client = client

    def execute(
        self,
        query: str,
        platform: Optional[str] = None,
    ) -> LlmExecuteResponse:
        """Execute a natural-language query via the LLM endpoint.

        Args:
            query: The natural-language query (e.g. "list my github repos").
            platform: Optional platform filter (e.g. "github", "slack").

        Returns:
            LlmExecuteResponse with execution results.

        Example::

            result = client.llm.execute("list my github repos")
            if result["ok"]:
                print(result["result"])
        """
        payload: Dict[str, Any] = {"query": query}
        if platform:
            payload["platform"] = platform

        return self._client._post("/api/llm", payload)

    def list_available_tools(
        self,
        platform: Optional[str] = None,
    ) -> LlmToolsResponse:
        """List available tools grouped by platform for LLM consumption.

        Args:
            platform: Optional platform filter (e.g. "github", "slack").

        Returns:
            LlmToolsResponse mapping platform slugs to PlatformTools.

        Example::

            resp = client.llm.list_available_tools()
            for platform, info in resp["platforms"].items():
                if info["connected"]:
                    print(f"{platform}: {len(info['tools'])} tools")
        """
        params: Dict[str, Any] = {}
        if platform:
            params["platform"] = platform

        return self._client._get("/api/llm/tools", params)
